go/ql/lib/semmle/go/frameworks/Twirp.qll

```diff
@@ -0,0 +1,175 @@
+      )
+    }
+  }
+
+  /**
+   * An interface type representing a Twirp service.
+   */
+  class ServiceInterface extends InterfaceType {
+    NamedType serviceInterface;
+
+    ServiceInterface() {
+      exists(TypeEntity te |
+        te.getType() = serviceInterface and
+        this instanceof InterfaceType and
+        serviceInterface.getUnderlyingType() = this and
+        te.getDeclaration().getLocation().getFile() instanceof ServicesGeneratedFile
+      )
+    }
+
+    /**
+     * Returns the name of the interface
+     */
+    override string getName() { result = serviceInterface.getName() }
+
+    /**
+     * Returns the named type on top of this interface type
+     */
+    NamedType getNamedType() { result = serviceInterface }
+  }
+
+  /**
+   * A Twirp client
+   */
+  class ServiceClient extends NamedType {
+    ServiceClient() {
+      exists(ServiceInterface i, PointerType p |
+        p.implements(i) and
+        this = p.getBaseType() and
+        this.getName().toLowerCase() = i.getName().toLowerCase() + ["protobuf", "json"] + "client"
+      )
+    }
+  }
+
+  /**
+   * A Twirp server
+   */
+  class ServiceServer extends NamedType {
+    ServiceServer() {
+      exists(ServiceInterface i |
+        this.implements(i) and
+        this.getName().toLowerCase() = i.getName().toLowerCase() + "server"
+      )
+    }
+  }
+
+  /**
+   * A Twirp function to construct a Client
+   */
+  class ClientConstructor extends Function {
+    ClientConstructor() {
+      exists(ServiceClient c |
+        this.getName().toLowerCase() = "new" + c.getName().toLowerCase() and
+        this.getParameter(0).getType().getName() = "string" and
+        this.getParameter(1).getType().getName() = "HTTPClient"
+      )
+    }
+  }
+
+  /**
+   * A Twirp function to construct a Server
+   * Its first argument should be an implementation of the service interface
+   */
+  class ServerConstructor extends Function {
+    ServerConstructor() {
+      exists(ServiceServer c, ServiceInterface i |
+        this.getName().toLowerCase() = "new" + c.getName().toLowerCase() and
+        this.getParameter(0).getType() = i.getNamedType()
+      )
+    }
+  }
+
+  /**
+   * An SSRF sink for the Client constructor
+   */
+  class ClientRequestUrlAsSink extends RequestForgery::Sink {
+    ClientRequestUrlAsSink() {
+      exists(DataFlow::CallNode call |
+        call.getArgument(0) = this and
+        call.getTarget() instanceof ClientConstructor
+      )
+    }
+
+    override DataFlow::Node getARequest() { none() }
+
+    override string getKind() { result = "URL" }
+  }
+
+  /**
+   * A service handler
+   */
+  class ServiceHandler extends Method {
+    ServiceHandler() {
+      exists(DataFlow::CallNode call, Type handlerType, ServiceInterface i |
+        call.getTarget() instanceof ServerConstructor and
+        call.getArgument(0).getType() = handlerType and
+        handlerType.implements(i) and
```

Comment:
This line is unnecessary. It is included in the call to `implements` two lines below.